controllers/pipelinerun_controller.go

```diff
@@ -0,0 +1,203 @@
+				UID:        pipelineRun.GetUID(),
+			}},
+		},
+		Spec: v1alpha1.BuildRunSpec{
+			BuildRef: &v1alpha1.BuildRef{
+				Name: buildName,
+			},
+		},
+	}
+	if err := r.Client.Create(ctx, &br); err != nil {
+		return "", err
+	}
+	return br.GetName(), nil
+}
+
+// triggerBuildsForPipelineRun create the BuildRun instances for the informed objects, and updates
+// the PipelineRun object labels to list the created objects.
+func (r *PipelineRunReconciler) triggerBuildsForPipelineRun(
+	ctx context.Context,
+	logger *logr.Logger,
+	pipelineRun *tknv1beta1.PipelineRun,
+	buildsToBeTriggered []inventory.SearchResult,
+) error {
+	var created []string
+	for _, build := range buildsToBeTriggered {
+		buildRunName, err := r.createBuildRun(ctx, pipelineRun, build.BuildName.Name)
+		if err != nil {
+			return err
+		}
+		created = append(created, buildRunName)
+	}
+	if len(created) == 0 {
+		return fmt.Errorf("no buildruns have been created for %q", pipelineRun.GetNamespacedName())
+	}
+	logger.V(0).Info("BuildRun(s) have been created",
+		"buildruns", created, "pipeline-run-name", pipelineRun.GetNamespacedName())
+
+	// adding a label to the PipelineRun object to identify the BuildRun(s) created for it, and also,
+	// to be able to filter out objects that have already triggered Builds afterwards
+	labels := pipelineRun.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[filter.BuildRunsCreatedKey] = strings.Join(created, ", ")
+	labels[filter.PipelineRunNameKey] = pipelineRun.GetName()
+	pipelineRun.SetLabels(labels)
+
+	return r.Client.Update(ctx, pipelineRun)
+}
+
+// Reconcile inspects the PipelineRun to extract the query parameters for the Build inventory search,
+// and at the end creates the BuildRun instance(s). Before firing the BuildRuns it inspects the
+// PipelineRun to assert the object is being referred by triggers and it's not part of a Custom-Task
+// Pipeline.
+func (r *PipelineRunReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var pipelineRun tknv1beta1.PipelineRun
+	if err := r.Get(ctx, req.NamespacedName, &pipelineRun); err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Unable to fetch PipelineRun")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !pipelineRun.ObjectMeta.DeletionTimestamp.IsZero() {
+		logger.V(0).Info("Marked for deletion")
+		return ctrl.Result{}, nil
+	}
+
+	if pipelineRun.Spec.PipelineRef == nil {
+		logger.V(0).Info("Skipping due to not having PipelineRef defined")
+		return ctrl.Result{}, nil
+	}
+
+	// if the label recording the original PipelineRun name, which triggered builds, matches the
+	// current object name it gets skipped from the rest of the syncing process
+	if !filter.PipelineRunNotSyncedAndNotCustomTask(&pipelineRun) {
+		logger.V(0).Info("Skipping due to either already synced or part of a custom-task")
+		return ctrl.Result{}, nil
+	}
```

Comment:
I think this one could also be moved to the predicate function ?

Beside that I tried to follow the code path. Are you here trying to prevent the re-evaluation of a PipelineRun for which BuildRuns were already created ? I was looking for something in that direction because a controller restart would reconcile all PipelineRuns and we need to somehow determine what we must trigger and what was already triggered. But not sure if your approach works. Different Builds can have triggers on the same Pipeline but with different status for example. Once the first status is reached, it will trigger BuildRun(s) and set the label. How will we then trigger additional BuildRuns once another status is reached for which other triggers exist ?

We maybe need to somehow remember which status of the PipelineRun we already synced?

Reply:
> I think this one could also be moved to the predicate function ?

Sure! Please consider the latest changes.
> 
> Beside that I tried to follow the code path. Are you here trying to prevent the re-evaluation of a PipelineRun for which BuildRuns were already created ?

Yes, that's exactly what the controller does.

> I was looking for something in that direction because a controller restart would reconcile all PipelineRuns and we need to somehow determine what we must trigger and what was already triggered. But not sure if your approach works.

Well, I've tested this approach it does prevent reprocessing `PipelineRun` instances.

> Different Builds can have triggers on the same Pipeline but with different status for example. Once the first status is reached, it will trigger BuildRun(s) and set the label. How will we then trigger additional BuildRuns once another status is reached for which other triggers exist ?

Indeed, that's a use-case the current approach does not cover. I think we might need to restructure the labels to point out which `Build` names have been triggered.
 
> We maybe need to somehow remember which status of the PipelineRun we already synced?

I think we might need to know which `Build`s have already being triggered by that `PipelineRun`.

Reply:
Please consider the latest changes, now we rely on a annotation to register previously triggered builds to skip repeated events.